test-common/src/test/scala/org/scalajs/testcommon/RPCCoreTest.scala

```diff
@@ -0,0 +1,101 @@
+package org.scalajs.testcommon
+
+import scala.concurrent._
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContext.Implicits.global
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import org.junit.Test
+import org.junit.Assert._
+
+class RPCCoreTest {
+  import RPCCoreTest._
+
+  lazy val x: TestRPC = new TestRPC(y)
+  lazy val y: TestRPC = new TestRPC(x)
+
+  object eps {
+    import RPCEndpoint._
+
+    val simple: EP[Unit, Unit] = ep[Unit, Unit](2)
+    val number: EP[Unit, Int] = ep[Unit, Int](3)
+  }
+
+  @Test
+  def simpleEndpoint: Unit = {
+    var called = false
+    x.attach(eps.simple)((_: Unit) => called = true)
+    Await.result(y.call(eps.simple)(()), atMost = 1.second)
+    assertTrue(called)
+  }
+
+  @Test
+  def multiplePendingCalls: Unit = {
+    val p = Promise[Int]
+
+    x.attachAsync(eps.number)(_ => p.future)
+
+    val futures = List.fill(20)(y.call(eps.number)(()))
+
+    p.success(1)
+
+    val results = Await.result(Future.sequence(futures), atMost = 10.second)
+    assertEquals(List.fill(20)(1), results)
+  }
+
+  private def assertRPCException(f: Future[_]): Throwable = {
+    try {
+      Await.result(f, atMost = 1.second)
+      fail("Expected exception")
+      throw new AssertionError("Shouldn't reach here")
+    } catch {
+      case e: RPCCore.RPCException =>
+        assertNotNull(e.getCause())
+        e.getCause()
+    }
+  }
+
+  @Test
+  def unboundEndpoint: Unit = {
+    assertRPCException(y.call(eps.simple)(()))
+  }
+
+  @Test
+  def remoteException: Unit = {
+    val msg0 = "My message for the outer exception"
+    val msg1 = "My message for the inner exception"
+    x.attach(eps.simple)(
+        (_: Unit) => throw new Exception(msg0, new Exception(msg1)))
```

Comment:
Both `simpleEndpoint` and `remoteException` do an `x.attach(eps.simple)`. How does that not fail?